core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2334,6 +2334,10 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
       }
 
+      // If this callback was executed after KafkaApis returned we will need to adjust the callback completion time here.
+      if (request.callbackRequestDequeTimeNanos.isDefined)
+        request.callbackRequestCompleteTimeNanos = Some(time.nanoseconds())
+
```

Comment:
Instead of doing this, I would put the request.callbackRequestCompleteTimeNanos = Some(time.nanoseconds()) back after the callback.fun() call and just update the place where we update metrics to take the current time if callbackRequestCompleteTimeNanos is empty.  Then we don't need chase all the code paths that we could go through before updating the metrics.

Reply:
Hmm -- so we should only update if callback is also defined. 
I'm also not sure about setting it twice, so we should only set after fun() if it is not already set.